File: src/modules/02-cell-site-identification/tower-source.ts
import type { CellTower, GeoZone } from '../../types/tower.js';

/**
 * TowerSource contract — requirement #2.
 *
 * Any module needing "all cell towers whose coverage falls inside this alert
 * zone" talks to this interface. Two adapters ship with TURANT:
 *
 *   - PostgisTowerSource : direct SQL against the C-DOT PostGIS database
 *   - HttpTowerSource    : C-DOT API gateway (awaiting the endpoint contract)
 *
 * Selection is purely config-driven (TOWER_SOURCE_MODE) — no code changes when
 * the real tower database arrives.
 */
export interface FindTowersOptions {
  /** Max rows returned (safety cap; real zones will usually exceed this and it is configurable). */
  limit?: number;
  /** Cancel the query (translated to a DB-side statement_timeout / fetch abort). */
  signal?: AbortSignal;
  /** CAP alert identifier for the shared latency trace (t1 is marked on it). */
  traceKey?: string;
}

export interface TowerSource {
  /** Stable name for logs/audit, e.g. "postgis" | "http". */
  readonly name: string;
  findTowersInZone(zone: GeoZone, options?: FindTowersOptions): Promise<CellTower[]>;
}

/** Convert a CAP-derived zone into the union of its GeoJSON geometries. */
export function zoneToGeoJson(zone: GeoZone): unknown[] {
  return zone.geometries.map((g) => {
    if (g.type === 'Polygon') {
      return { type: 'Polygon', coordinates: g.coordinates };
    }
    return {
      type: 'Circle',
      center: [g.center!.lng, g.center!.lat],
      radiusMeters: g.radiusMeters,
    };
  });
}

File: src/main/java/com/turant/cellsite/TowerSource.java
package com.turant.cellsite;

import com.turant.types.tower.CellTower;
import com.turant.types.tower.GeoZone;

import java.util.List;
import java.util.concurrent.CompletableFuture;

/**
 * TowerSource contract - requirement #2.
 * 
 * Any module needing "all cell towers whose coverage falls inside this alert zone"
 * talks to this interface.
 * 
 * Migrated from TypeScript Module 02 tower-source.ts
 */
public interface TowerSource {
    
    /**
     * Stable name for logs/audit, e.g. "postgis" | "http" | "memory"
     */
    String getName();
    
    /**
     * Find all towers whose coverage intersects the given zone.
     *
     * @param zone Alert geographic zone
     * @param options Search options (limit, timeout, tracing)
     * @return List of matching cell towers
     */
    CompletableFuture<List<CellTower>> findTowersInZone(GeoZone zone, FindTowersOptions options);
    
    /**
     * Options for tower search.
     */
    class FindTowersOptions {
        private Integer limit;
        private Long timeoutMs;
        private String traceKey;
        
        public Integer getLimit() {
            return limit;
        }
        
        public FindTowersOptions setLimit(Integer limit) {
            this.limit = limit;
            return this;
        }
        
        public Long getTimeoutMs() {
            return timeoutMs;
        }
        
        public FindTowersOptions setTimeoutMs(Long timeoutMs) {
            this.timeoutMs = timeoutMs;
            return this;
        }
        
        public String getTraceKey() {
            return traceKey;
        }
        
        public FindTowersOptions setTraceKey(String traceKey) {
            this.traceKey = traceKey;
            return this;
        }
        
        public static FindTowersOptions defaults() {
            return new FindTowersOptions();
        }
    }
}
